Python/_warnings.c

```diff
@@ -431,6 +431,45 @@ call_show_warning(PyObject *category, PyObject *text, PyObject *message,
     return -1;
 }
 
+static int
+tb_displayline(PyObject *f, PyObject *filename, int lineno, PyObject *name)
+{
+    int err;
+    PyObject *line;
+
+    if (filename == NULL || name == NULL)
+        return -1;
+    line = PyUnicode_FromFormat("  File \"%U\", line %d, in %U\n",
+                                filename, lineno, name);
+    if (line == NULL)
+        return -1;
+    err = PyFile_WriteObject(line, f, Py_PRINT_RAW);
+    Py_DECREF(line);
+    if (err != 0)
+        return err;
+    /* ignore errors since we can't report them, can we? */
+    if (_Py_DisplaySourceLine(f, filename, lineno, 4))
+        PyErr_Clear();
+    return err;
+}
+
+
+static void
+print_frames(PyFrameObject *frame)
+{
+    PyObject *f_stderr;
+    f_stderr = _PySys_GetObjectId(&PyId_stderr);
+
+    while (frame != NULL){
+        tb_displayline(
+            f_stderr,
+            frame->f_code->co_filename,
+            PyCode_Addr2Line(frame->f_code, frame->f_lasti),
```

Comment:
PyFrame_GetLineNumber() must be used here.

nitpick: I would prefer to compute it before calling tb_displayline():

int lineno = PyFrame_GetLineNumber(frame);
tb_displayline(...);